### PR TITLE
Conditional strings for Whiskers.

### LINKED ISSUE #8733: Add a "nonempty" conditional to Whiskers
In several whiskers templates, we have patters of the form
```
<hasParams> let <params> := </hasParams> f()
```

or

```
function <functionName>() <arrow> <retVariales> {
```

Both could be simplified by a "some param is nonempty" conditional instead of adding another param manually. We could use `#param` for that:

```
<?#params> let <params> := </#params> f()
```

or

```
function <functionName>() <?#retVariables> -> <retVariales> </#retVariables> {
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #8733

### CODE REVIEW

**Review by mijovic — DISMISSED** [review]
Looks good to me

File: docs/contributing.rst [hunk 1]
```diff
@@ -356,6 +356,11 @@ by as many concatenations of its contents as there were sets of variables suppli
 each time replacing any ``<inner>`` items by their respective value. Top-level variables can also be used
 inside such areas.
 
+There are also conditionals of the form ``<?name>...<!name>...</name>``, where template replacements
+continues recursively either in the first or the second segment depending on the value of the boolean
```
Comment (leonardoalt) [inline docs/contributing.rst]:
```suggestion
continue recursively either in the first or the second segment depending on the value of the Boolean
```

**Review by leonardoalt — CHANGES_REQUESTED** [review]
The syntax `+<b>` doesn't seem to be documented?

**Comment by chriseth** [comment]
I think I'll change this to use `+`

File: test/libsolutil/Whiskers.cpp [hunk 2]
```diff
@@ -114,6 +114,19 @@ BOOST_AUTO_TEST_CASE(conditional_plus_list)
...
+{
+	string templ = "<?+b>+<b><!+b>-</+b>";
+	BOOST_CHECK_EQUAL(Whiskers(templ)("b", "abc").render(), "+abc");
+	BOOST_CHECK_EQUAL(Whiskers(templ)("b", "").render(), "-");
+}
+
+BOOST_AUTO_TEST_CASE(string_as_conditional_wrong)
+{
+	string templ = "<?+b>+<b></b>";
+	BOOST_CHECK_EQUAL(Whiskers(templ)("b", "abc").render(), "<?+b>+abc</b>");
```
Comment (leonardoalt) [inline test/libsolutil/Whiskers.cpp:127]:
```suggestion
	BOOST_CHECK_EQUAL(Whiskers(templ)("b", "abc").render(), "+abc");
```
?

**Review by leonardoalt — CHANGES_REQUESTED** [review]

File: test/libsolutil/Whiskers.cpp (hunk 2 above)
Comment (chriseth) [reply test/libsolutil/Whiskers.cpp:127]: Take a closer look ;)

Comment (leonardoalt) [reply test/libsolutil/Whiskers.cpp:127]: Ah ;)

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**